### ISSUE TITLE
pubsub: context deadline exceeded emitted in iterator.sendModAck

### ISSUE THREAD

**Issue by hongalex** [googleapis/google-cloud-go#3060]
**Client**

Pub/Sub

**Environment**

macOS

**Go Environment**

go 1.15

**Description**

`ModifyAckDeadline` calls use a context that has a deadline of 2 seconds. Normally, any context deadline exceeded errors are sent from the server which attaches an appropriate `DeadlineExceeded` code, which the pubsub library treats as a non-fatal error.

**Expected behavior**

Any `context deadline exceeded` errors for ModAcks are non-fatal.

**Actual behavior**

Sometimes, fatal errors are bubbled up to the user (via the iterator) and requires a restart of the `sub.Receive` call.

**Reproduction**

Start streaming pull via `Receive` and hold messages in the client without acking immediately (give approximately 10 second delay). Turn off internet connection to the application (e.g. disable WiFi on laptop). After waiting approximately 30 minutes, turn on WiFi and allow acks/modacks RPCs in the iterator to start again. Using an annotated of the library that handles error wrapping, one or two `ModAck` calls will fail and I saw results: 

```
modack error: context deadline exceeded
ctx is: context.Background.WithDeadline(2020-10-16 01:40:14.9839 -0700 PDT m=+1495.611993088 [-987.861µs])```
```
Notice the modack error does not contain the `rpc error` prefix, which indicates the RPC never reached the server, or the error is coming from elsewhere. The `gax.Sleep` call can be ruled out, even though it also uses the same context, because non-nil errors fall through to the `codes.DeadlineExceeded` case.

**Comment by shangsunset** [comment]
Hi, we are also getting `context deadline exceeded` error from pubsub `Receive`. This started to happen after we enabled GCP workload identity. We suspect that gke-metadata-server might be interfering with the network between google pubsub client and server? This goes away when turning off workload identity or switching from synchronous pull to async pull as it doesn't have a timeout.

**Comment by hongalex** [comment]
Thanks for reporting @shangsunset. Could you describe how often you encounter `context deadline exceeded` errors after enabling workload identity? Do you have a working reproduction by any chance, or how long does it take before these errors are surfaced?

**Comment by shangsunset** [comment]
Hey @hongalex, the error shows up pretty consistently when workload identity is on. Our application would exit from the error, then fails the readiness and liveness probs and cause the pods to restart over and over. Usually, this starts to happen within minutes after a fresh deployment.

**Comment by hongalex** [comment]
Actually, one more clarification. Are you saying you have issues with *only* the synchronous pull client, i.e. `sub.ReceiveSettings.Synchronous = true`, when workload identity is turned on? And when you're using asynchronous pull and workload identity, everything works fine?

**Comment by shangsunset** [comment]
Correct. We chose synchronous pull because we don't get any dupes. I also want to point out that the older version of pubsub client(1.3.x) works fine with workload identity and `sub.ReceiveSettings.Synchronous = true`

**Comment by hongalex** [comment]
Still waiting on a reproduction, but assuming workload identity is an issue, it's likely related to some connection changes made since v1.3.x. Will report back with an update by next week.

**Comment by lordnynex** [comment]
+1 on this issue. Receive random `context deadline exceeded` from `Subscription.Receive()`. I don't have much other info because I can't reliably reproduce this and at the point an error is returned, it's a fatal error for me. 

```
	cloud.google.com/go/pubsub v1.10.0
```

**Comment by santowijaya28** [comment]
hi, got same issue as well. We are using `Subscription.Receive()`.  But somehow, it will resulting on context deadline exceeded and we need to reload our service. We are using this version of pubsub 
```
	cloud.google.com/go/pubsub v1.5.0
```

**Comment by hongalex** [comment]
Sorry, just about to look at this again.

@lordnynex Was this also with workload identity turned on?

@santowijaya28 I actually mitigated this in v1.8.3 and later. Could you try updating to the latest version and see if it still happens?

**Comment by meredithslota** [comment]
Closing as we believe this was fixed in v1.8.3 but please add more details and repro steps if this doesn't fix your issue.
